File: applications/mobile-rn/src/domain/bacm/sceneGeometry.ts
import { AppConfig } from "@/config/AppConfig";
import type { MessageLayer, MessageScene } from "./scene";

export type LayerRect = { x: number; y: number; w: number; h: number };

export function layerBounds(layer: MessageLayer): LayerRect {
  switch (layer.type) {
    case "text":
    case "photo":
      return { x: layer.x, y: layer.y, w: layer.w, h: layer.h };
    case "icon":
      return { x: layer.x, y: layer.y, w: layer.size, h: layer.size };
    default:
      return { x: 0, y: 0, w: 0, h: 0 };
  }
}

export function hitTestLayer(
  layers: MessageLayer[],
  px: number,
  py: number,
): MessageLayer | null {
  for (let i = layers.length - 1; i >= 0; i--) {
    const layer = layers[i];
    if (layer.hidden) continue;
    const b = layerBounds(layer);
    if (px >= b.x && px < b.x + b.w && py >= b.y && py < b.y + b.h) return layer;
  }
  return null;
}

export function clampLayerPosition(
  x: number,
  y: number,
  layer: MessageLayer,
): { x: number; y: number } {
  const b = layerBounds(layer);
  const maxX = Math.max(0, AppConfig.MSG_WIDTH - b.w);
  const maxY = Math.max(0, AppConfig.MSG_HEIGHT - b.h);
  return {
    x: Math.max(0, Math.min(maxX, x)),
    y: Math.max(0, Math.min(maxY, y)),
  };
}

export function nextLayerId(layers: MessageLayer[], prefix: string): string {
  const nums = layers
    .filter((l) => l.id.startsWith(prefix))
    .map((l) => parseInt(l.id.slice(prefix.length), 10))
    .filter((n) => !Number.isNaN(n));
  const max = nums.length > 0 ? Math.max(...nums) : 0;
  return `${prefix}${max + 1}`;
}

export function sceneVisualKey(scene: MessageScene): string {
  let out = `${scene.bgColor}|${scene.bgType}|${scene.bgImageUri ?? ""}|`;
  for (const l of scene.layers) {
    out += `${l.id}:${l.type}:${l.hidden}:${l.text}:${l.ref}:${l.w}:${l.h}:${l.size}:${l.color}:${l.fontSize}:${l.fontId}:${l.align}:${l.anim}:${l.imageUri ?? ""};`;
  }
  return out;
}

export function scenePositionKey(scene: MessageScene): string {
  return scene.layers.map((l) => `${l.id}:${l.x},${l.y}`).join(";");
}

File: applications/mobile/shared/src/commonMain/kotlin/com/tchy/boiteacoeur/domain/bacm/SceneGeometry.kt
package com.tchy.boiteacoeur.domain.bacm

import com.tchy.boiteacoeur.AppConfig
import com.tchy.boiteacoeur.data.model.MessageLayer

data class LayerRect(val x: Int, val y: Int, val w: Int, val h: Int)

fun layerBounds(layer: MessageLayer): LayerRect = when (layer.type) {
    "text" -> LayerRect(layer.x, layer.y, layer.w, layer.h)
    "photo" -> LayerRect(layer.x, layer.y, layer.w, layer.h)
    "icon" -> LayerRect(layer.x, layer.y, layer.size, layer.size)
    else -> LayerRect(0, 0, 0, 0)
}

fun hitTestLayer(layers: List<MessageLayer>, px: Int, py: Int): MessageLayer? {
    for (i in layers.indices.reversed()) {
        val layer = layers[i]
        if (layer.hidden) continue
        val b = layerBounds(layer)
        if (px in b.x until b.x + b.w && py in b.y until b.y + b.h) return layer
    }
    return null
}

fun clampLayerPosition(x: Int, y: Int, layer: MessageLayer): Pair<Int, Int> {
    val b = layerBounds(layer)
    val maxX = (AppConfig.MSG_WIDTH - b.w).coerceAtLeast(0)
    val maxY = (AppConfig.MSG_HEIGHT - b.h).coerceAtLeast(0)
    return x.coerceIn(0, maxX) to y.coerceIn(0, maxY)
}

fun nextLayerId(layers: List<MessageLayer>, prefix: String): String {
    val nums = layers.filter { it.id.startsWith(prefix) }.mapNotNull { layer ->
        layer.id.removePrefix(prefix).toIntOrNull()
    }
    return "$prefix${(nums.maxOrNull() ?: 0) + 1}"
}

fun emojiRefToFolder(ref: String): String {
    val raw = ref.removePrefix("emoji:").trim()
    return raw
        .replace(":", "_")
        .replace(Regex("[^a-zA-Z0-9_]"), "_")
        .removePrefix("emoji_")
}

fun sceneVisualKey(scene: com.tchy.boiteacoeur.data.model.MessageScene): String = buildString {
    append(scene.bgColor)
    append('|')
    append(scene.bgType)
    append('|')
    scene.layers.forEach { layer ->
        append(layer.id)
        append(':')
        append(layer.type)
        append(':')
        append(layer.hidden)
        append(':')
        append(layer.text)
        append(':')
        append(layer.ref)
        append(':')
        append(layer.w)
        append(':')
        append(layer.h)
        append(':')
        append(layer.size)
        append(':')
        append(layer.color)
        append(':')
        append(layer.fontSize)
        append(':')
        append(layer.align)
        append(':')
        append(layer.anim)
        append(';')
    }
}

fun scenePositionKey(scene: com.tchy.boiteacoeur.data.model.MessageScene): String =
    scene.layers.joinToString(";") { "${it.id}:${it.x},${it.y}" }
